app/src/main/java/com/concordium/wallet/data/preferences/Preferences.kt

```diff
@@ -15,9 +19,145 @@ open class Preferences {
     interface Listener {
         fun onChange()
     }
+    companion object{
+        const val SHARED_PREFERENCES_ARE_ENCRYPTED ="encrypted_shared_preference"
+    }
 
     constructor(context: Context, preferenceName: String, preferenceMode: Int) {
-        sharedPreferences = context.getSharedPreferences(preferenceName, preferenceMode)
+        val authPreferences = initializeEncryptedSharedPreferences(context, SharedPreferencesKeys.PREF_FILE_AUTH)
+        if(authPreferences.getBoolean(SHARED_PREFERENCES_ARE_ENCRYPTED, false)){
+            sharedPreferences = initializeEncryptedSharedPreferences(context, preferenceName)
+        }else{
+            sharedPreferences =
+                if(migratePreferencesSuccess(context, preferenceName, preferenceMode, authPreferences)){
+                    initializeEncryptedSharedPreferences(context, preferenceName)
+                }else{
+                    getSharedPreferences(context, preferenceName, preferenceMode)
+                }
+        }
+    }
+
+    /**
+     * Returns the required SharedPreferences
+     * @return [SharedPreferences] *if* unencrypted data is still present
+     *
+     * *else* an instance of encrypted [SharedPreferences]
+     */
+    private fun getSharedPreferences(
+        context: Context,
+        preferenceName: String,
+        preferenceMode: Int
+    ): SharedPreferences {
+        val unEncryptedSharedPreferences = context.getSharedPreferences(preferenceName, preferenceMode)
+
+        return if(unEncryptedSharedPreferences.all.isEmpty()) {
+            initializeEncryptedSharedPreferences(context, preferenceName)
+        }else{
+            unEncryptedSharedPreferences
+        }
+    }
+
+    /**
+     * Loops through all of the [SharedPreferences] and attempts to save them to [EncryptedSharedPreferences]
+     * @return *true* if the [SharedPreferences] don't need migrating
+     *
+     * *true* if [SharedPreferences] is successfully migrated and the old data cleared
+     *
+     * *false* if the migrations fail or is not cleared successfully
+     */
+    private fun migratePreferencesSuccess(
+        mContext: Context,
+        preferenceName: String,
+        preferenceMode: Int,
+        authPreferences: SharedPreferences): Boolean{
+        var allPreferencesAreMigrated = true
+        var continueWithEncryptedSharedPreference = true
```

Comment:
I don't understand `continueWithEncryptedSharedPreference`. What continuation does this control? There are no checks based on it, it's just the value returned from the function.

Reply:
If the migration of the unencrypted SharedPreferences fails for some reason for the required preference name, then `continueWithEncryptedSharedPreference` will be false and the app should continue and use the unencrypted SharedPreference. It tells the app which SharedPreference to use, should it use the unencrypted shared preferences or continue with the encrypted ones.

Reply:
> There's something odd about this mixed logic where we try to migrate everything while reading one thing. It seems unnecessarily complex and will keep "migrating" things that are already migrated all the time.

` if(unencryptedPreference.all.isNotEmpty()){`

It will not keep migrating things because it first checks if the old unencrypted SharedPreference is empty. We migrate everything while reading one thing because some of the SharedPreferences are rarely used and they will stay unencrypted until they are called,  which may never happen.

>I also don't think we need to handle the case where the migration succeeded partially too gracefully: Either it migrated completely and we use the encrypted ones exclusively or something failed and we use the unencrypted ones exclusively.

This would result with all of the shared preferences remaining unencrypted on the device if a single shared preferences keeps failing to migrate.


>I would much prefer that we fire a migration on app startup (if needed), and that if it succeeds it sets the success flag. When reading the prefs, we just read that flag to determine what kinds of prefs to read from.

The migration is fired on app startup, when we first engage a SharedPreference to check if the user is logged in, and a success flag for the migration is set and read:

` if(allPreferencesAreMigrated){
            authPreferences.edit().putBoolean(SHARED_PREFERENCES_ARE_ENCRYPTED, true).commit()
        }`  

`if(authPreferences.getBoolean(SHARED_PREFERENCES_ARE_ENCRYPTED, false)){
            initializeEncryptedSharedPreferences(context, preferenceName)
        }`

>If it's an option to block usage of the app by the migration being successful (like we ended up doing for the seed phrase migration?), I would do that and then only expect to find the encrypted ones afterwards.

By the time we get to the Auth screen Shared Preferences are already been initialised and used to check the login status and other stuff, migrating them while they are used can result in unexpected behaviour.  So its best to attempt to migrate them when they are first called.

Reply:
> In the Preference we need to know which SharedPreferences to use, the old unencrypted or the encrypted ones. There must be a check to decide between them and this check must happen every time the Preference is initialised. First a global check is done: `if(authPreferences.getBoolean(SHARED_PREFERENCES_ARE_ENCRYPTED, false)){ ` and if this is true then we won't even reach the migration part.

With my suggestion you would only check the `SHARED_PREFERENCES_ARE_ENCRYPTED` flag and load the appropriate variant based on that. I don't think it's that different from what you did, I would just do the actual migration once, somewhere else and consider it entirely succeeded or entirely failed.

> Regarding the showing of dialogs, popups and warnings to the user and different approaches to the task, this is a reappearing problem where the team has not discussed and made an decision on how the problem should be solved so that I as a developer have a clear understanding what needs to be done.
> 
> The discussion is left for after the task is done, this slows down the development and increases frustration for everyone on the team.

I think that part of taking on a task is also exploring the solution space and seeking consensus on the best approach. I agree that it would be helpful with a more structured process on how we do this, particularly when it comes to deciding how to involve the user.

I don't think that discussions on deeper changes in review is necessarily a bad thing. I just shows that there is more need for seeking consensus. Though it is of course desirable to address disagreements as early as possible.

In this case it would probably have been a good idea to write a small design document outlining the approach and send it around the team. Whether to do that before or after the initial implementation is your call as the task owner. I would usually do the work first because it gives me valuable insight, even if I end up having to change it.

I think this is a good topic for retrospection @ali-concordium .